### ISSUE TITLE
Assignment: persistent or not

### ISSUE THREAD

**Issue by NicoLaval** [VTL-Community/VTL-Community#24]
We were wondering with @hadrienk what is the real difference between `<-` (persistent assignment) and `:=` (non-persistent assignment).

Does that mean datasets assigned with non-persistent assignment operator must not be available outside of the calculation scope?
Non-persistent assignment is created to temporarily define intermediate datasets which don't matter then? (for readability, code sharing, performances)

```
DS := DS1 + DS2;
A <- round(DS);
B <- mod(DS, 2);
```

Instead of

```
A <- round(DS1 + DS2);
B <- mod(DS1 + DS2, 2);
```

Have you any real world use case?

**Comment by antonio-olleros** [comment]
Hi, we have lots of thems!
We have several transformation schemes with dependencies between transformations, and most of the transformations are only for calculation purposes, but they are not needed as output of the transformation scheme. So, for us it is very important that the user makes explicit what is to be kept after the run time, and what can be dropped after the runtime, because it is only used as an intermediate step.

For instance, in this TS:
<img width="547" alt="image" src="https://github.com/VTL-Community/VTL-Community/assets/65357445/8c19b7a8-ec1f-4a64-8c81-2902614738d6">

You can see that we have 22 input datasets to produce 10 output datasets, and we have 250 intermediate datasets that are required to get to the output, but are not needed as output.

Hope this clarifies!

**Comment by NicoLaval** [comment]
Great, ty @antonio-olleros

**Comment by NicoLaval** [comment]
Hello,

I reopen the issue because I would like to be sure that we all share the same point of view.

Considering this script:
```
DS := DS1 + DS2;
A <- round(DS);
B <- mod(DS, 2);
```

Input bindings are `DS1` and `DS2`.
After script execution, what would return the engine as output bindings?

- only persistent variables: `A`, `B`?
- input bindings and persistent variables: `A`, `B`, `DS1`, `DS2`?
- all: `A`, `B`, `DS`, `DS1`, `DS2`?

@antonio-olleros, considering the 2 first solutions, you would not be able to list intermediate DS in your IHM (does that mean we have to provide several execution mode, with variability scope?)

At this point, I don't think VTL docs impose the behavior, do you think it's a good idea to define clearly that point in the docs? Or leave free each engine to return a scope or an other?

At least, hope to have a maximum of advice about what has to be output bindings.

**Comment by antonio-olleros** [comment]
Hi!

In our implementation, we provide a parameter, to let executers decide. By default, the execution via API returns only the persistent (i.e. option 1: A and B). If the user decides so, returns DS, A and B (new option). We don't return the input, since it was provided by the system that is expecting the response.

Then, we have what we call the vtlPlayground, which provides functionalities for user to write and test VTL. In that app, users are always seeing everything (option 3). But this is because we want them to be able to easily get all the information, so that they can better write VTL.

To summarise, the different operators are needed, because users should be able to decide what they want to keep and what they don't. What is their impact, I think it should be for the implementations to decide.

Hope this is helpful!

**Comment by NicoLaval** [comment]
Ty @antonio-olleros 
I think it's a good way to do it.

**Comment by linardian** [comment]
I think this issue can be closed
